### ISSUE TITLE
App crash when using BackgroundGeolocation.onGeofencesChange

### ISSUE THREAD

**Issue by joaolh** [transistorsoft/flutter_background_geolocation#65]
## Your Environment
* Plugin version: 1.0.7
* Platform: Android
* OS version: 9
* Device manufacturer / model: Google Pixel
* Flutter info (`flutter doctor`):
* Plugin config:         
foregroundService: true,
        notificationText: "Text",
        notificationTitle: "Title",
        desiredAccuracy: bg.Config.DESIRED_ACCURACY_HIGH,
        distanceFilter: 10.0,
        geofenceProximityRadius: 1600,
        stopOnTerminate: false,
        startOnBoot: true,
        debug: true,
        logLevel: bg.Config.LOG_LEVEL_VERBOSE,
        reset: true
```dart <-- Syntax highlighting: DO NOT REMOVE -->

## Expected Behavior
<!--- Tell us what should happen -->

## Actual Behavior
<!--- Tell us what happens instead -->

## Steps to Reproduce
<!--- reproduce this issue; include code to reproduce, if relevant -->
1. Code is the demo example code from the github. Add a couple geofences.
2. Add bg.BackgroundGeolocation.onGeofencesChange(_onGeofencesChange);
3. The _onGeofencesChange has nothing on it, just a print statement.
4.

## Context
<!--- What were you trying to do? -->

## Debug logs
<!-- include iOS / Android logs
- ios XCode logs,
- use #getLog #emailLog methods (@see docs)
- Android: $ adb logcat -s TSLocationManager
-->
<details>
	<summary>Logs</summary>

``` <!-- syntax-highligting:  DO NOT REMOVE -->
Launching lib/main.dart on Pixel in debug mode...
Built build/app/outputs/apk/debug/app-debug.apk.
I/OpenGLRenderer(13001): Initialized EGL, version 1.4
D/OpenGLRenderer(13001): Swap behavior 1
D/TSLocationManager(13001): [c.t.f.b.streams.StreamHandler onListen] location
D/TSLocationManager(13001): [c.t.f.b.streams.StreamHandler onListen] motionchange
D/TSLocationManager(13001): [c.t.f.b.streams.StreamHandler onListen] activitychange
D/TSLocationManager(13001): [c.t.f.b.streams.StreamHandler onListen] providerchange
D/TSLocationManager(13001): [c.t.f.b.streams.StreamHandler onListen] connectivitychange
D/TSLocationManager(13001): [c.t.f.b.streams.StreamHandler onListen] geofenceschange
D/TSLocationManager(13001): [c.t.f.b.streams.StreamHandler onListen] geofence
I/TSLocationManager(13001): [c.t.locationmanager.data.a.a a]
I/TSLocationManager(13001):   ✅  Home
I/TSLocationManager(13001): [c.t.locationmanager.data.a.a a]
I/TSLocationManager(13001):   ✅  Work
I/TSLocationManager(13001): [c.t.l.u.BackgroundTaskManager onStartJob] ⏳ startBackgroundTask: 1
D/TSLocationManager(13001): [c.t.l.g.TSGeofenceManager$b run] evaluation buffer timer elapsed
E/TSLocationManager(13001): [c.t.l.l.TSLocationManager getLastLocation]
E/TSLocationManager(13001):   ‼️  Failed to get last location: com.google.android.gms.tasks.zzu@454231e
I/flutter (13001): [addGeofences] success
D/TSLocationManager(13001): [c.t.l.adapter.TSConfig c] ℹ️   Persist config
D/TSLocationManager(13001): [c.t.locationmanager.util.b a]
D/TSLocationManager(13001):   ℹ️  LocationAuthorization: Permission granted
I/TSLocationManager(13001): - Enable: true → true, trackingMode: 1
D/TSLocationManager(13001): [c.t.l.a.BackgroundGeolocation ready] LocationPermission :true
I/TSLocationManager(13001): [c.t.l.service.HeartbeatService a]
I/TSLocationManager(13001):   🔴  Stop heartbeat
I/TSLocationManager(13001): [c.t.l.u.BackgroundTaskManager onStartJob] ⏳ startBackgroundTask: 3
I/TSLocationManager(13001): [c.t.l.s.ActivityRecognitionService a]
I/TSLocationManager(13001):   🎾  Start motion-activity updates
D/TSLocationManager(13001): [c.t.locationmanager.util.b a]
D/TSLocationManager(13001):   ℹ️  LocationAuthorization: Permission granted
I/flutter (13001): [ProviderChangeEvent enabled:true, status: 3, network: true, gps: true]
I/TSLocationManager(13001): [c.t.l.g.TSGeofenceManager start]
I/TSLocationManager(13001):   🎾  Start monitoring geofences
I/TSLocationManager(13001): [c.t.l.u.BackgroundTaskManager onStartJob] ⏳ startBackgroundTask: 2
D/TSLocationManager(13001): [c.t.locationmanager.util.b a]
D/TSLocationManager(13001):   ℹ️  LocationAuthorization: Permission granted
D/TSLocationManager(13001): [c.t.l.http.HttpService startMonitoringConnectivityChanges]
D/TSLocationManager(13001):   🎾  Start monitoring connectivity changes
D/TSLocationManager(13001): [c.t.locationmanager.device.a c]
D/TSLocationManager(13001):   🎾  Start monitoring powersave changes
D/TSLocationManager(13001): [c.t.l.s.ActivityRecognitionService a]
D/TSLocationManager(13001):   🚘 ️DetectedActivity [type=STILL, confidence=100]
D/TSLocationManager(13001): [c.t.l.s.ActivityRecognitionService onDestroy]
D/TSLocationManager(13001):   🔴  ActivityRecognitionService destroyed
I/flutter (13001): ----------------------- [ready] -
I/flutter (13001): [activitychange] - [ActivityChangeEvent still (100%)]
I/TSLocationManager(13001): [c.t.l.s.ActivityRecognitionService a]
I/TSLocationManager(13001): ╔═════════════════════════════════════════════
I/TSLocationManager(13001): ║ Motion Transition Result
I/TSLocationManager(13001): ╠═════════════════════════════════════════════
I/TSLocationManager(13001): ╟─ 🎾  ENTER: still
I/TSLocationManager(13001): ╚═════════════════════════════════════════════
D/EventBus(13001): No subscribers registered for event class java.util.Collections$UnmodifiableRandomAccessList
D/EventBus(13001): No subscribers registered for event class org.greenrobot.eventbus.NoSubscriberEvent
I/flutter (13001): [activitychange] - [ActivityChangeEvent still (100%)]
D/TSLocationManager(13001): [c.t.l.s.ActivityRecognitionService onDestroy]
D/TSLocationManager(13001):   🔴  ActivityRecognitionService destroyed
D/TSLocationManager(13001): [c.t.l.g.TSGeofenceManager a]
D/TSLocationManager(13001): ╔═════════════════════════════════════════════
D/TSLocationManager(13001): ║ TSGeofenceManager found 1/2 within 1600 meters
D/TSLocationManager(13001): ╠═════════════════════════════════════════════
D/TSLocationManager(13001): ╟─ 🎾  Home
D/TSLocationManager(13001): ╚═════════════════════════════════════════════
D/AndroidRuntime(13001): Shutting down VM
E/AndroidRuntime(13001): FATAL EXCEPTION: main
E/AndroidRuntime(13001): Process: com.transistorsoft.flutterbackgroundgeolocationexample, PID: 13001
E/AndroidRuntime(13001): java.lang.NullPointerException: Attempt to invoke virtual method 'java.util.Iterator org.json.JSONObject.keys()' on a null object reference
E/AndroidRuntime(13001): 	at com.transistorsoft.locationmanager.util.Util.toMap(Unknown Source:5)
E/AndroidRuntime(13001): 	at com.transistorsoft.locationmanager.geofence.TSGeofence.toMap(Unknown Source:110)
E/AndroidRuntime(13001): 	at com.transistorsoft.locationmanager.event.GeofencesChangeEvent.toMap(Unknown Source:53)
E/AndroidRuntime(13001): 	at com.transistorsoft.flutter.backgroundgeolocation.streams.GeofencesChangeStreamHandler.onGeofencesChange(GeofencesChangeStreamHandler.java:23)
E/AndroidRuntime(13001): 	at com.transistorsoft.locationmanager.geofence.TSGeofenceManager.a(Unknown Source:24)
E/AndroidRuntime(13001): 	at com.transistorsoft.locationmanager.geofence.TSGeofenceManager.a(Unknown Source:546)
E/AndroidRuntime(13001): 	at com.transistorsoft.locationmanager.geofence.TSGeofenceManager.setLocation(Unknown Source:221)
E/AndroidRuntime(13001): 	at com.transistorsoft.locationmanager.location.TSLocationManager.onSingleLocationResult(Unknown Source:122)
E/AndroidRuntime(13001): 	at com.transistorsoft.locationmanager.location.SingleLocationRequest$2.onLocationResult(Unknown Source:164)
E/AndroidRuntime(13001): 	at com.google.android.gms.internal.location.zzau.notifyListener(Unknown Source:4)
E/AndroidRuntime(13001): 	at com.google.android.gms.common.api.internal.ListenerHolder.notifyListenerInternal(Unknown Source:17)
E/AndroidRuntime(13001): 	at com.google.android.gms.common.api.internal.ListenerHolder$zaa.handleMessage(Unknown Source:5)
E/AndroidRuntime(13001): 	at android.os.Handler.dispatchMessage(Handler.java:106)
E/AndroidRuntime(13001): 	at android.os.Looper.loop(Looper.java:193)
E/AndroidRuntime(13001): 	at android.app.ActivityThread.main(ActivityThread.java:6680)
E/AndroidRuntime(13001): 	at java.lang.reflect.Method.invoke(Native Method)
E/AndroidRuntime(13001): 	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:493)
E/AndroidRuntime(13001): 	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:858)
E/TSLocationManager(13001): [c.t.l.a.BackgroundGeolocation$e uncaughtException]
E/TSLocationManager(13001):   ‼️  Uncaught Exception: Attempt to invoke virtual method 'java.util.Iterator org.json.JSONObject.keys()' on a null object reference
E/TSLocationManager(13001): {"activityRecognitionInterval":10000,"allowIdenticalLocations":false,"autoSync":true,"autoSyncThreshold":0,"batchSync":false,"debug":true,"deferTime":0,"desiredAccuracy":-1,"desiredOdometerAccuracy":100,"disableElasticity":false,"disableStopDetection":false,"distanceFilter":10,"elasticityMultiplier":1,"enableHeadless":false,"enableTimestampMeta":false,"extras":{},"fastestLocationUpdateInterval":-1,"forceReloadOnBoot":false,"forceReloadOnGeofence":false,"forceReloadOnHeartbeat":false,"forceReloadOnLocationChange":false,"forceReloadOnMotionChange":false,"forceReloadOnSchedule":false,"foregroundService":true,"geofenceInitialTriggerEntry":true,"geofenceModeHighAccuracy":false,"geofenceProximityRadius":1600,"geofenceTemplate":"","headers":{},"headlessJobService":"com.transistorsoft.flutter.backgroundgeolocation.HeadlessTask","heartbeatInterval":-1,"httpRootProperty":"location","httpTimeout":60000,"isMoving":false,"locationTemplate":"","locationTimeout":60,"locationUpdateInterval":1000,"locationsOrderDirection":"ASC","logLevel":5,"logMaxDays":3,"maxBatchSize":-1,"maxDaysToPersist":1,"maxRecordsToPersist":-1,"method":"POST","minimumActivityRecognitionConfidence":75,"notificationColor":"","notificationLargeIcon":"","notificationPriority":0,"notificationSmallIcon":"","notificationText":"Text","notificationTitle":"Title","params":{},"persist":true,"persistMode":2,"schedule":[],"speedJumpFilter":300,"startOnBoot":true,"stationaryRadius":25,"stopAfterElapsedMinutes":0,"stopOnStationary":false,"stopOnTerminate":false,"stopTimeout":5,"triggerActivities":"in_vehicle, on_bicycle, on_foot, running, walking","url":"","useSignificantChangesOnly":false,"enabled":true,"schedulerEnabled":false,"trackingMode":1,"odometer":0,"isFirstBoot":false}
E/TSLocationManager(13001): java.lang.NullPointerException: Attempt to invoke virtual method 'java.util.Iterator org.json.JSONObject.keys()' on a null object reference
E/TSLocationManager(13001): 	at com.transistorsoft.locationmanager.util.Util.toMap(Unknown Source:5)
E/TSLocationManager(13001): 	at com.transistorsoft.locationmanager.geofence.TSGeofence.toMap(Unknown Source:110)
E/TSLocationManager(13001): 	at com.transistorsoft.locationmanager.event.GeofencesChangeEvent.toMap(Unknown Source:53)
E/TSLocationManager(13001): 	at
com.transistorsoft.flutter.backgroundgeolocation.streams.GeofencesChangeStreamHandler.onGeofencesChange(GeofencesChangeStreamHandler.java:23)
E/TSLocationManager(13001): 	at com.transistorsoft.locationmanager.geofence.TSGeofenceManager.a(Unknown Source:24)
E/TSLocationManager(13001): 	at com.transistorsoft.locationmanager.geofence.TSGeofenceManager.a(Unknown Source:546)
E/TSLocationManager(13001): 	at com.transistorsoft.locationmanager.geofence.TSGeofenceManager.setLocation(Unknown Source:221)
E/TSLocationManager(13001): 	at com.transistorsoft.locationmanager.location.TSLocationManager.onSingleLocationResult(Unknown Source:122)
E/TSLocationManager(13001): 	at com.transistorsoft.locationmanager.location.SingleLocationRequest$2.onLocationResult(Unknown Source:164)
E/TSLocationManager(13001): 	at com.google.android.gms.internal.location.zzau.notifyListener(Unknown Source:4)
E/TSLocationManager(13001): 	at com.google.android.gms.common.api.internal.ListenerHolder.notifyListenerInternal(Unknown Source:17)
E/TSLocationManager(13001): 	at com.google.android.gms.common.api.internal.ListenerHolder$zaa.handleMessage(Unknown Source:5)
E/TSLocationManager(13001): 	at android.os.Handler.dispatchMessage(Handler.java:106)
E/TSLocationManager(13001): 	at android.os.Looper.loop(Looper.java:193)
E/TSLocationManager(13001): 	at android.app.ActivityThread.main(ActivityThread.java:6680)
E/TSLocationManager(13001): 	at java.lang.reflect.Method.invoke(Native Method)
E/TSLocationManager(13001): 	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:493)
E/TSLocationManager(13001): 	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:858)
I/Process (13001): Sending signal. PID: 13001 SIG: 9
Lost connection to device.
Exited (sigterm)```

```

</details>

**Comment by christocracy** [comment]
Are you executing #removeGeofences somewhere?

**Comment by christocracy** [comment]
Show me the Geofences you're adding.

**Comment by christocracy** [comment]
I see the problem.  Provide an empty `{}` to `Geofence.extras`.  This will be fixed in the next release.

```dart
bg.BackgroundGeolocation.addGeofence(bg.Geofence(
        ...        
        extras: {}    // <-- provide an empty `{}`
    ))
```

**Comment by joaolh** [comment]
Yeap, this might be it. I just found out that the "BackgroundGeolocation.geofences" also crashes. But this time I was able to see that "extras" must be present to prevent the crash. So, extras technically is mandatory field, unless an exception handling is added to the method.

**Comment by christocracy** [comment]
#extras is supposed to be optional.  It’s a bug, already fixed in my local version.  It will be fixed in the next release.

**Comment by christocracy** [comment]
For now, simply provide an empty {} to extras and carry on.

**Comment by christocracy** [comment]
Fixed in 1.0.8
